### PR TITLE
implement BufReader::peek

### PR DESCRIPTION
**Pull request by lolbinarycat**
Part of https://github.com/rust-lang/rust/issues/128405

### CODE REVIEW

File: library/std/src/io/buffered/bufreader.rs [hunk 1]
```diff
@@ -94,6 +94,32 @@ impl<R: Read> BufReader<R> {
...
+    /// let mut s = String::new();
+    /// rdr.read_to_string(&mut s).unwrap();
+    /// assert_eq!(&s, "hello");
+    /// ```
+    #[unstable(feature = "bufreader_peek", issue = "128405")]
+    pub fn peek(&mut self, n: usize) -> io::Result<&[u8]> {
+        assert!(n < self.capacity());
+        while n > self.buf.buffer().len() {
+            self.buf.read_more(&mut self.inner)?;
+        }
```
Comment (the8472) [inline library/std/src/io/buffered/bufreader.rs:120]: `buf` can have the current data sitting near the end, so the overall capacity of the buffer minus the length is not the remaining tail capacity.

So even if `n < capacity` it doesn't mean that amount can be read. So a backshift can be necessary.

You should also add internal tests covering that case.

File: library/std/src/io/buffered/bufreader/buffer.rs [hunk 2]
```diff
@@ -97,6 +97,25 @@ impl Buffer {
...
+        unsafe {
+            buf.set_init(old_init);
+        }
+        reader.read_buf(buf.unfilled())?;
+        self.filled += buf.len();
+        self.initialized += buf.init_len() - old_init;
+        Ok(())
+    }
+
+    pub fn backshift(&mut self) {
```
Comment (fbstj) [inline library/std/src/io/buffered/bufreader/buffer.rs:112]: would be nice to have some docs for these fn? thanks

Comment (lolbinarycat) [reply library/std/src/io/buffered/bufreader/buffer.rs:114]: done

**Comment by Mark-Simulacrum** [comment]
r=me with commits squashed

**Comment by lolbinarycat** [comment]
r=@Mark-Simulacrum

**Comment by the8472** [comment]
r=@Mark-Simulacrum

**Pull request merged**